<?xml version="1.0"?>
<launch>
  <arg name="camera_id"  default="8"/>
  <arg name="color_width" default="1920"/>
  <arg name="color_high" default="1080"/>
  <arg name="charuco_row" default="7"/>
  <arg name="charuco_col" default="5"/>
  <arg name="square_length" default="0.036"/>
  <arg name="marker_length" default="0.0244"/>
  <!-- Set the eye mode (see README for more info) -->
  <arg name="camera_parent_frame"/>
  <arg name="marker_parent_frame"/>

  <!-- The ARUCO marker properties -->
  <arg name="markerid"              default="582"/>
  <arg name="markersize"            default="0.127"/>    <!-- in m -->
  <arg name="sample_rate"           default="2"/>

  <!-- The camera topics and frame to use for calibration -->
  <arg name="camera" doc="camera namespace"/>
  <arg name="image"                 default="$(arg camera)/image_rect_color"/>
  <arg name="camera_info"           default="$(arg camera)/camera_info"/>
  <arg name="camera_frame"          default="$(arg camera)_optical_frame"/>
  <arg name="marker_frame"          default="/hand_eye/$(arg camera)/aruco_marker_frame"/>

  <!-- leave empty and the pose will be published wrt param parent_name -->
  <arg name="reference_frame"       default="$(arg camera_frame)"/>

  <!-- Publish frames while calibrating -->
  <arg name="publish_tf" default="true"/>

  <!-- Ask to commit after each sample -->
  <arg name="interactive" default="true"/>

  <!-- If the connector node should be spawned in a new window -->
  <arg name="separate_window" default="false"/>
  <arg unless="$(arg separate_window)" name="LAUNCH_PREFIX" value=""/>
  <arg if="$(arg separate_window)" name="LAUNCH_PREFIX" value="xterm -e"/>

  <!-- Transform from the camera base link to the optical link (only necessary
       if you want aruco_hand_eye to compute the transform for you) -->
  <arg name="xyz_optical_base" default="[0.0, 0.0, 0.0]"/>
  <arg name="rpy_optical_base" default="[0.0, 0.0, 0.0]"/>

  <group ns="$(arg camera)">

    <!-- Track the aruco target in a given camera -->
   
    <!-- Run the VISP solver to determine the extrinsic parameters -->
    <node name="hand_eye_solver"
      pkg="visp_hand2eye_calibration"
      type="visp_hand2eye_calibration_calibrator">
      <remap from="/compute_effector_camera_quick" to="compute_effector_camera_quick"/>
    </node>

    <!-- Connect the aruco tracker to the solver -->
    <node name="hand_eye_connector"
      launch-prefix="$(arg LAUNCH_PREFIX)"
      pkg="hand_eye"
      type="calibrate.py"
      output="screen">
      <param name="marker_id"          value="$(arg markerid)"/>
      <param name="reference_frame"    value="$(arg reference_frame)"/>
      <param name="camera_frame"       value="$(arg camera_frame)"/>
      <param name="marker_frame"       value="$(arg marker_frame)" />
      <param name="sample_rate" value="$(arg sample_rate)"/>
      <param name="camera_parent_frame" value="$(arg camera_parent_frame)"/>
      <param name="marker_parent_frame" value="$(arg marker_parent_frame)"/>
      <param name="publish_tf" value="$(arg publish_tf)"/>
      <param name="tf_suffix" value=""/>
      <param name="interactive" value="$(arg interactive)"/>
      <rosparam param="xyz_optical_base" subst_value="true">$(arg xyz_optical_base)</rosparam>
      <rosparam param="rpy_optical_base" subst_value="true">$(arg rpy_optical_base)</rosparam>
    </node>
    
    <node name="charuco_tracker"
      launch-prefix="$(arg LAUNCH_PREFIX)"
      pkg="hand_eye"
      type="CharucoPosture.py"
      output="screen">
      <param name="color_width" value="$(arg color_width)"/>
      <param name="color_high" value="$(arg color_high)"/>
      <param name="camera_id" value="$(arg camera_id)"/>
      <param name="charuco_row"   value="$(arg charuco_row)"/>
      <param name="charuco_col"   value="$(arg charuco_col)"/>
      <param name="square_length" value="$(arg square_length)"/>
      <param name="marker_length" value="$(arg marker_length)"/>
    </node>
    
  </group>

</launch>
